File: data/benchmark/src/benchmark/parsers/codex_cli.py
"""Codex CLI JSONL transcript parser.

Port of packages/ingest/src/parsers/codex-cli.ts
"""

from __future__ import annotations

import json
import re

from .types import ParsedSession, ParsedTurn, ToolUseRecord, ToolResultRecord


def parse_codex_cli_file(file_path: str) -> list[ParsedSession]:
    """Parse a Codex CLI JSONL transcript file into sessions."""
    with open(file_path, encoding="utf-8") as f:
        raw = f.read()
    return parse_codex_cli_jsonl(raw, file_path)


def parse_codex_cli_jsonl(content: str, file_path: str = "") -> list[ParsedSession]:
    """Parse Codex CLI JSONL content string into sessions."""
    session_id: str | None = None
    cwd: str | None = None
    model_id: str | None = None
    started_at: str | None = None
    ended_at: str | None = None
    turns: list[ParsedTurn] = []

    for line in content.splitlines():
        if not line.strip():
            continue
        try:
            parsed = json.loads(line)
        except json.JSONDecodeError:
            continue

        ts = parsed.get("timestamp", "")
        if ts:
            if not started_at or ts < started_at:
                started_at = ts
            if not ended_at or ts > ended_at:
                ended_at = ts

        line_type = parsed.get("type")
        payload = parsed.get("payload", {})
        if not isinstance(payload, dict):
            payload = {}

        if line_type == "session_meta":
            session_id = payload.get("id")
            cwd = payload.get("cwd")
            model_id = payload.get("model") or payload.get("model_provider")
            if payload.get("timestamp"):
                started_at = payload["timestamp"]
            elif parsed.get("timestamp"):
                started_at = parsed["timestamp"]

        elif line_type == "turn_context":
            if payload.get("model"):
                model_id = payload["model"]
            if payload.get("cwd"):
                cwd = payload["cwd"]

        elif line_type == "response_item":
            payload_type = payload.get("type")

            if payload_type == "message":
                # Skip developer/system messages
                role = payload.get("role")
                if role in ("developer", "system"):
                    continue
                mapped_role = "assistant" if role == "assistant" else "user"

                text_content = ""
                p_content = payload.get("content")
                if isinstance(p_content, list):
                    for block in p_content:
                        if isinstance(block, dict) and block.get("type") in ("output_text", "input_text") and block.get("text"):
                            text_content += block["text"] + "\n"

                if text_content.strip():
                    turns.append(ParsedTurn(
                        role=mapped_role,
                        text_content=text_content.strip(),
                        tool_uses=[],
                        tool_results=[],
                        timestamp=ts,
                    ))

            elif payload_type == "function_call":
                tool_name = payload.get("name", "shell")
                input_data: dict = {}
                args = payload.get("arguments")
                if args:
                    try:
                        input_data = json.loads(args)
                    except (json.JSONDecodeError, TypeError):
                        input_data = {"raw": args}

                call_id = payload.get("call_id", "")
                turns.append(ParsedTurn(
                    role="assistant",
                    text_content="",
                    tool_uses=[ToolUseRecord(tool_name=tool_name, input=input_data, id=call_id)],
                    tool_results=[],
                    timestamp=ts,
                ))

            elif payload_type == "custom_tool_call":
                tool_name = payload.get("name", "custom_tool")
                input_data = {"raw": payload.get("input", "")}
                call_id = payload.get("call_id", "")
                turns.append(ParsedTurn(
                    role="assistant",
                    text_content="",
                    tool_uses=[ToolUseRecord(tool_name=tool_name, input=input_data, id=call_id)],
                    tool_results=[],
                    timestamp=ts,
                ))

            elif payload_type in ("function_call_output", "custom_tool_call_output"):
                call_id = payload.get("call_id", "")
                output = payload.get("output", "")
                is_error = "error" in output.lower() or "failed" in output.lower()
                turns.append(ParsedTurn(
                    role="user",
                    text_content="",
                    tool_uses=[],
                    tool_results=[ToolResultRecord(
                        tool_use_id=call_id,
                        content=output[:1000],
                        is_error=is_error,
                    )],
                    timestamp=ts,
                ))

        elif line_type == "event_msg":
            event = parsed.get("event")
            if event in ("user_message",) or payload.get("type") == "user_message":
                text = (
                    payload.get("message")
                    or payload.get("text")
                    or (payload.get("content", [{}])[0].get("text") if isinstance(payload.get("content"), list) else "")
                    or ""
                )
                if text:
                    turns.append(ParsedTurn(role="user", text_content=text, timestamp=ts))
            elif event in ("agent_message",) or payload.get("type") == "agent_message":
                text = (
                    payload.get("message")
                    or payload.get("text")
                    or (payload.get("content", [{}])[0].get("text") if isinstance(payload.get("content"), list) else "")
                    or ""
                )
                if text:
                    turns.append(ParsedTurn(role="assistant", text_content=text, timestamp=ts))

    if not session_id:
        safe = re.sub(r"[^a-z0-9]", "_", file_path, flags=re.IGNORECASE)[-40:]
        session_id = f"codex_{safe}"

    if not started_at:
        from datetime import datetime, timezone

        started_at = datetime.now(timezone.utc).isoformat()

    sessions: list[ParsedSession] = []
    if turns:
        sessions.append(ParsedSession(
            id=session_id,
            platform="codex_cli",
            model_id=model_id,
            cwd=cwd,
            git_branch=None,
            started_at=started_at,
            ended_at=ended_at,
            turns=turns,
            file_path=file_path,
        ))
    return sessions

File: packages/ingest/src/parsers/codex-cli.ts
import { readFileSync } from "node:fs";
import type { ParsedSession, ParsedTurn, ToolUseRecord, ToolResultRecord, SourcePlatform } from "@sathergate/vendor-observatory-shared";

interface CodexContent {
  type: string;
  text?: string;
}

interface CodexPayload {
  type?: string;
  role?: string;
  content?: CodexContent[];
  name?: string;
  arguments?: string;
  call_id?: string;
  output?: string;
  text?: string;
  message?: string;
  input?: string; // custom_tool_call input
  // session_meta / turn_context payload fields
  id?: string;
  timestamp?: string;
  cwd?: string;
  model?: string;
  model_provider?: string;
}

interface CodexLine {
  type: string;
  timestamp?: string;
  payload?: CodexPayload;
  event?: string;
}

/**
 * Parse a Codex CLI JSONL transcript file into sessions.
 */
export function parseCodexCliFile(filePath: string): ParsedSession[] {
  const content = readFileSync(filePath, "utf-8");
  const lines = content.split("\n").filter(Boolean);

  let sessionId: string | null = null;
  let cwd: string | null = null;
  let modelId: string | null = null;
  let startedAt: string | null = null;
  let endedAt: string | null = null;
  const turns: ParsedTurn[] = [];

  for (const line of lines) {
    let parsed: CodexLine;
    try {
      parsed = JSON.parse(line) as CodexLine;
    } catch {
      continue;
    }

    const ts = parsed.timestamp ?? new Date().toISOString();
    if (!startedAt || ts < startedAt) startedAt = ts;
    if (!endedAt || ts > endedAt) endedAt = ts;

    if (parsed.type === "session_meta") {
      const p = parsed.payload;
      sessionId = p?.id ?? null;
      cwd = p?.cwd ?? null;
      modelId = p?.model ?? p?.model_provider ?? null;
      if (p?.timestamp) startedAt = p.timestamp;
      else if (parsed.timestamp) startedAt = parsed.timestamp;
    } else if (parsed.type === "turn_context") {
      const p = parsed.payload;
      if (p?.model) modelId = p.model;
      if (p?.cwd) cwd = p.cwd;
    } else if (parsed.type === "response_item") {
      const payload = parsed.payload;
      if (!payload) continue;

      if (payload.type === "message") {
        // Skip developer/system messages (sandbox permissions, instructions, etc.)
        if (payload.role === "developer" || payload.role === "system") continue;

        const role = payload.role === "assistant" ? "assistant" : "user";
        let textContent = "";

        if (payload.content && Array.isArray(payload.content)) {
          for (const block of payload.content) {
            if ((block.type === "output_text" || block.type === "input_text") && block.text) {
              textContent += block.text + "\n";
            }
          }
        }

        if (textContent.trim()) {
          turns.push({
            role: role as "user" | "assistant",
            textContent: textContent.trim(),
            toolUses: [],
            toolResults: [],
            timestamp: ts,
          });
        }
      } else if (payload.type === "function_call") {
        const toolName = payload.name ?? "shell";
        let input: Record<string, unknown> = {};
        try {
          if (payload.arguments) {
            input = JSON.parse(payload.arguments) as Record<string, unknown>;
          }
        } catch {
          input = { raw: payload.arguments };
        }

        const callId = payload.call_id ?? `call_${Date.now()}`;
        const toolUse: ToolUseRecord = { toolName, input, id: callId };

        turns.push({
          role: "assistant",
          textContent: "",
          toolUses: [toolUse],
          toolResults: [],
          timestamp: ts,
        });
      } else if (payload.type === "custom_tool_call") {
        // Custom tool calls (e.g., apply_patch) — treat like function_call
        const toolName = payload.name ?? "custom_tool";
        const input: Record<string, unknown> = { raw: payload.input ?? "" };
        const callId = payload.call_id ?? `call_${Date.now()}`;
        const toolUse: ToolUseRecord = { toolName, input, id: callId };

        turns.push({
          role: "assistant",
          textContent: "",
          toolUses: [toolUse],
          toolResults: [],
          timestamp: ts,
        });
      } else if (payload.type === "function_call_output" || payload.type === "custom_tool_call_output") {
        const callId = payload.call_id ?? "";
        const output = payload.output ?? "";
        const isError = output.toLowerCase().includes("error") || output.toLowerCase().includes("failed");
        const toolResult: ToolResultRecord = {
          toolUseId: callId,
          content: output.slice(0, 1000),
          isError,
        };

        turns.push({
          role: "user",
          textContent: "",
          toolUses: [],
          toolResults: [toolResult],
          timestamp: ts,
        });
      }
    } else if (parsed.type === "event_msg") {
      const payload = parsed.payload;
      if (!payload) continue;

      if (parsed.event === "user_message" || payload.type === "user_message") {
        const text = payload.message ?? payload.text ?? (payload.content?.[0]?.text) ?? "";
        if (text) {
          turns.push({
            role: "user",
            textContent: text,
            toolUses: [],
            toolResults: [],
            timestamp: ts,
          });
        }
      } else if (parsed.event === "agent_message" || payload.type === "agent_message") {
        const text = payload.message ?? payload.text ?? (payload.content?.[0]?.text) ?? "";
        if (text) {
          turns.push({
            role: "assistant",
            textContent: text,
            toolUses: [],
            toolResults: [],
            timestamp: ts,
          });
        }
      }
    }
  }

  if (!sessionId) {
    sessionId = `codex_${filePath.replace(/[^a-z0-9]/gi, "_").slice(-40)}`;
  }

  if (!startedAt) {
    startedAt = new Date().toISOString();
  }

  const sessions: ParsedSession[] = [];
  if (turns.length > 0) {
    sessions.push({
      id: sessionId,
      platform: "codex_cli" as SourcePlatform,
      modelId,
      cwd,
      gitBranch: null,
      startedAt,
      endedAt,
      turns,
      filePath,
    });
  }

  return sessions;
}
